File: src/java/1_SOLIDDesignPrinciples/5_DependencyInversion/DIP.java
package com.activemesa.solid.dip;

import org.javatuples.Triplet;

import java.util.ArrayList;
import java.util.List;
import java.util.Objects;
import java.util.stream.Collectors;

// A. High-level modules should not depend on low-level modules. 
// Both should depend on abstractions.

// B. Abstractions should not depend on details. 
// Details should depend on abstractions.

enum Relationship
{
  PARENT,
  CHILD,
  SIBLING
}

class Person
{
  public String name;
  // dob etc.


  public Person(String name) {
    this.name = name;
  }
}

interface RelationshipBrowser
{
  List<Person> findAllChildrenOf(String name);
}

class Relationships implements RelationshipBrowser
{
  public List<Person> findAllChildrenOf(String name) {

    return relations.stream()
      .filter(x -> Objects.equals(x.getValue0().name, name)
              && x.getValue1() == Relationship.PARENT)
      .map(Triplet::getValue2)
      .collect(Collectors.toList());
  }

  // Triplet class requires javatuples
  private List<Triplet<Person, Relationship, Person>> relations =
    new ArrayList<>();

  public List<Triplet<Person, Relationship, Person>> getRelations() {
    return relations;
  }

  public void addParentAndChild(Person parent, Person child)
  {
    relations.add(new Triplet<>(parent, Relationship.PARENT, child));
    relations.add(new Triplet<>(child, Relationship.CHILD, parent));
  }
}

class Research
{
  public Research(Relationships relationships)
  {
    // high-level: find all of john's children
    List<Triplet<Person, Relationship, Person>> relations = relationships.getRelations();
    relations.stream()
      .filter(x -> x.getValue0().name.equals("John")
              && x.getValue1() == Relationship.PARENT)
      .forEach(ch -> System.out.println("John has a child called " + ch.getValue2().name));
  }

  public Research(RelationshipBrowser browser)
  {
    List<Person> children = browser.findAllChildrenOf("John");
    for (Person child : children)
      System.out.println("John has a child called " + child.name);
  }
}

class DIPDemo
{
  public static void main(String[] args)
  {
    Person parent = new Person("John");
    Person child1 = new Person("Chris");
    Person child2 = new Person("Matt");

    // low-level module
    Relationships relationships = new Relationships();
    relationships.addParentAndChild(parent, child1);
    relationships.addParentAndChild(parent, child2);

    new Research(relationships);
  }
}

File: src/swift/1_SOLIDDesignPrinciples/5_DependencyInversion/DIP.swift
import Foundation

// hl modules should not depend on low-level; both should depend on abstractions
// abstractions should not depend on details; details should depend on abstractions

enum Relationship
{
  case parent
  case child
  case sibling
}

class Person
{
  var name = ""
  // dob, etc
  init(_ name: String)
  {
    self.name = name
  }
}

protocol RelationshipBrowser
{
  func findAllChildrenOf(_ name: String) -> [Person]
}

class Relationships : RelationshipBrowser
{
  var relations = [(Person, Relationship, Person)]()

  func addParentAndChild(_ p: Person, _ c: Person)
  {
    relations.append((p, Relationship.parent, c))
    relations.append((c, Relationship.child, p))
  }

  func findAllChildrenOf(_ name: String) -> [Person]
  {
    return relations.filter({$0.name == name && $1 == Relationship.parent && $2 != nil})
                    .map({$2})
  }
}

class Research
{
  init(_ relationships: Relationships)
  {
    // high-level: find all of job's children
    let relations = relationships.relations
    for r in relations where r.0.name == "John" && r.1 == Relationship.parent {
      print("John has a child called \(r.2.name)")
    }
  }

  init(_ browser: RelationshipBrowser)
  {
    for p in browser.findAllChildrenOf("John")
    {
      print("John has a child called \(p.name)")
    }
  }
}

func main()
{
  let parent = Person("John")
  let child1 = Person("Chris")
  let child2 = Person("Matt")

  var relationships = Relationships()
  relationships.addParentAndChild(parent, child1)
  relationships.addParentAndChild(parent, child2)

  let _ = Research(relationships)
}

main()